File: java/javafind/lib/src/main/java/javafind/FindConfig.java
package javafind;

import java.nio.file.Paths;

public class FindConfig {
    public static final String DEFAULT_FILE_TYPES_PATH = "/filetypes.json";
    public static final String DEFAULT_FIND_OPTIONS_PATH = "/findoptions.json";
    public static final String DEFAULT_XFIND_CONFIG_DIR =
            Paths.get(System.getProperty("user.home"), ".config", "xfind").toString();

    private final String fileTypesPath;
    private final String findOptionsPath;
    private final String defaultFindSettingsPath;

    public FindConfig() {
        this.fileTypesPath = DEFAULT_FILE_TYPES_PATH;
        this.findOptionsPath = DEFAULT_FIND_OPTIONS_PATH;
        var xFindConfigDir = getXFindConfigDir();
        this.defaultFindSettingsPath = Paths.get(xFindConfigDir, "settings.json").toString();
    }

    public String getFileTypesPath() {
        return fileTypesPath;
    }

    public String getFindOptionsPath() {
        return findOptionsPath;
    }

    public String getXFindConfigDir() {
        var xfindConfigDir = System.getenv("XFIND_CONFIG_DIR");
        if (xfindConfigDir == null || xfindConfigDir.isEmpty()) {
            xfindConfigDir = DEFAULT_XFIND_CONFIG_DIR;
        }
        return xfindConfigDir;
    }

    public String getDefaultFindSettingsPath() {
        return defaultFindSettingsPath;
    }
}

File: cpp/cppfind/src/FindConfig.cpp
#include "FindConfig.h"

namespace cppfind {
    std::string get_xfind_path() {
        const char* xf_path = std::getenv("XFIND_PATH");
        if (xf_path == nullptr) {
            const char* home = std::getenv("HOME");
            if (home == nullptr) {
                // TODO: throw exception?
                return "";
            }
            // TODO: make this cross-platform
            return std::string(home) + XFIND_REL_PATH;
        }
        return xf_path;
    }

    std::string get_file_types_path() {
        const std::string xf_path = get_xfind_path();
        return xf_path + "/" + FILE_TYPES_REL_PATH;
    }

    std::string get_find_options_path() {
        const std::string xf_path = get_xfind_path();
        return xf_path + "/" + FIND_OPTIONS_REL_PATH;
    }

    std::string get_xfind_config_dir() {
        const char* xf_config_dir = std::getenv("XFIND_CONFIG_DIR");
        if (xf_config_dir == nullptr) {
            const char* home = std::getenv("HOME");
            if (home == nullptr) {
                // TODO: throw exception?
                return "";
            }
            // TODO: make this cross-platform
            return std::string(home) + "/" + XFIND_CONFIG_REL_DIR;
        }
        return xf_config_dir;
    }

    std::string get_default_find_settings_path() {
        const std::string xf_config_dir = get_xfind_config_dir();
        return xf_config_dir + "/settings.json";
    }

    FindConfig::FindConfig() :
        m_xfind_path(get_xfind_path()),
        m_file_types_path(get_file_types_path()),
        m_find_options_path(get_find_options_path()),
        m_default_find_settings_path(get_default_find_settings_path()) {
    }

    std::string FindConfig::xfind_path() const {
        return m_xfind_path;
    }

    std::string FindConfig::file_types_path() const {
        return m_file_types_path;
    }

    std::string FindConfig::find_options_path() const {
        return m_find_options_path;
    }

    std::string FindConfig::default_find_settings_path() const {
        return m_default_find_settings_path;
    }
}
